realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -99,6 +99,11 @@
     final ConcurrentHashMap<WeakReference<RealmObjectProxy>, Object> realmObjects =
             new ConcurrentHashMap<WeakReference<RealmObjectProxy>, Object>();
 
+    // List of onSuccess callbacks from async transactions. We need to track all callbacks as notifying listeners might
+    // be delayed due to the presence of async queries. This can mean that multiple async transactions can complete
+    // before we are ready to notify all of them.
+    private final List<Runnable> pendingOnSuccessAsyncTransactionCallbacks = new ArrayList<Runnable>();
```

Comment:
Callbacks are added through the looper queue, so if a job is canceled they will never be sent in the first place. This queue is part of the HandlerController (which is GC'ed if the Realm is closed) and will be emptied on any other REALM_CHANGED event as well. So there shouldn't be any risk of leaking.